S2VX.Game/Editor/ToolState/SelectToolState.cs

```diff
@@ -192,41 +217,42 @@ public class SelectToolState : S2VXToolState {
             return true;
         }
 
-        private double GetClosestTickTime(double gameTime) {
-            var numTicks = Story.BPM / 60f * (Editor.Track.Length / NotesTimeline.SecondsToMS) * Editor.NotesTimeline.Divisor;
-            var timeBetweenTicks = Editor.Track.Length / numTicks;
-            var leftOffset = (gameTime - Story.Offset) % timeBetweenTicks;
-            var rightOffset = timeBetweenTicks - leftOffset;
-            return gameTime + (leftOffset <= rightOffset ? -leftOffset : rightOffset);
-        }
-
-        private double GetGameTimeAtMouse(Vector2 localMousePos) {
-            var noteSelectionIndicators = Editor.NotesTimeline.NoteSelectionIndicators;
-            var mousePosX = ToSpaceOfOtherDrawable(ToLocalSpace(localMousePos), noteSelectionIndicators).X;
-            // temp until NoteTimeline Scroll on drag is implemented
-            mousePosX = Math.Clamp(mousePosX, 0, noteSelectionIndicators.DrawWidth);
-            var relativeMousePosX = mousePosX / noteSelectionIndicators.DrawWidth;
-            var gameTimeDeltaFromMiddle = (relativeMousePosX - 0.5f) * Editor.NotesTimeline.SectionLength * NotesTimeline.SecondsToMS;
-            return Time.Current + gameTimeDeltaFromMiddle;
-        }
-
         public override void OnToolDrag(DragEvent e) {
             if (!DelayDrag) {
                 switch (ToDrag) {
                     case SelectToolDragState.DragHoldNoteEndTime: {
```

Comment:
Since this file's getting pretty big with nested switch cases, a refactoring suggestion would be to make `DragState` its own state class that implements `DragStart`, `Drag`, `DragEnd` functions.

Reply:
So this would be another ToolState that we'd switch to when Dragging?

Reply:
No, it would be implemented similarly to `ToolState`, but it's only maintained in the `SelectToolState` as a substate, if that makes sense. For instance, some pseudocode:

```csharp
interface ISelectToolDragState {
    OnToolDragStart();
    OnToolDrag();
    OnToolDragEnd();
}

class DragTimelineNote : ISelectToolDragState {
    DragTimelineNote(SelectToolState toolState) { ... }
    OnToolDragStart() { ... }
    OnToolDrag() { ... }
    OnToolDragEnd() { ... }
}

class AnotherDragState : ISelectToolDragState {
    DragTimelineNote(SelectToolState toolState) { ... }
    OnToolDragStart() { ... }
    OnToolDrag() { ... }
    OnToolDragEnd() { ... }
}

class SelectToolState {
    ISelectToolDragState DragState;

    OnToolDragStart() {
        if (someCondition) {
            DragState = new DragTimelineNote(this);
        } else if (someOtherCondition) {
            DragState = new AnotherDragState(this);
        } else {
            ...
        }
        DragState.DragStart();
    }

    OnToolDrag() {
        DragState.Drag();
    }

    OnToolDragEnd() {
        DragState.DragEnd();
    }
}
```